### PR TITLE
 [#871, #1065] Add item charges recovery using recovery formula

### LINKED ISSUE #871: Add a system for automatic recovery of charges for items like wands and other inventory that recover X uses per day with a dice roll.
Originally in GitLab by @thingscouldgetdicey1

a way to track charges for items like wands inventory items that require to recharge on a long rest and roll a die. to give the amount of charges to that item. if it could be linked to the long rest mechanic and roll in the chat as a chat card as to how many charges they get. it would be a good system to track the charges involved and that would be unique to foundry. if you could specify a field that says this item has X charges per day with a roll field. for example wand of fireballs reads as. 

 Wand of Fireballs

Requires Attunement by a Spellcaster

This wand has 7 Charges. While holding it, you can use an action to expend 1 or more of its Charges to cast the Fireball spell (save DC 15) from it. For 1 charge, you cast the 3rd-level version of the spell. You can increase the spell slot level by one for each additional charge you expend.

The wand regains 1d6 + 1 expended Charges daily at dawn. If you expend the wand's last charge, roll a d20. On a 1, the wand crumbles into ashes and is destroyed

so what would be useful is that if on a long rest or short rest depending on the items properties set by the dm or the item in a compendium. if that button was clicked that it would auto roll the charges for the players and add the stipulation of the destroyed on a 1 mechanic that was an optional setting in the settings menu. or within the item menu whichever is implemented best. like a box that you can click to say destroy on a 1 on next rest if all charges expended.

so process is as follows

Item is created. 

Item has a marked field specifiying it is an item that requires charges per dice roll with the formula dice + modifier (number or other field such as ability scores etc)

Item has a marked field that specifies that if all charges are expended chance of item breaking is introduced on a 1. 

Item has a field that specifies if charges come back on short or long rest linked to the roll mechanic.

DM or player rolls short or long rest. charges are dealt with and player just expends charges upon use. maybe with a pop up warning on their last charge (optional) with a message that reads you feel the power fading from your (item name) or maybe in the chat a chat card with that stating only one charge left. 

let me know what you think

### LINKED ISSUE #1065: New day on short rest is inconsistent between rest variants
Currently the short rest dialog is set up to display the "New Day" prompt as long as the rest variant isn't set to _epic_, but the value of that prompt is only returned when the rest variant is _gritty_ which leads to nothing happening if the checkbox is selected on normal rest mode. There are two ways to resolve this issue:

1. New day is possible on normal short rest, therefor the value should passed to the short rest process
2. New day is impossible on normal short rest, therefor the prompt should be hidden unless the rest variant is set to _gritty_

### PR DESCRIPTION
**Pull request by arbron**
Adds `system.uses.recoveryFormula` to the `Item` type which allows the items with `charges` to be automatically recharged based on a roll when a new day occurs.

Recharge Formula

Recharge Roll

This formula is only used if an item has the `charges` setting for uses and a new day has past.

- [ ] Update the `Items (SRD)` compendium to add these formulas

Resolves #871
Resolves #1065

### CODE REVIEW

**Comment by arbron** [comment]
> When is a new day considered to have occurred?

Whenever you click the "New Day" button in the short or long rest dialog.

File: module/documents/actor/actor.mjs [hunk 1]
```diff
@@ -1716,6 +1726,31 @@ export default class Actor5e extends Actor {
...
+          const count = Math.min(item.system.uses.max - item.system.uses.value, roll.total);
+          const isMax = (item.system.uses.value + count) === item.system.uses.max;
+          const locKey = `DND5E.ItemRecoveryRoll${isMax ? "Max" : ""}`;
+          updates.push({_id: item.id, "system.uses.value": item.system.uses.value + count});
+          roll.options.flavor = game.i18n.format(locKey, { name: item.name, count });
+          rolls.push(roll);
+          await roll.toMessage({
+            user: game.user.id,
+            speaker: { actor: this, alias: this.name }
+          });
```
Comment (Fyorl) [inline module/documents/actor/actor.mjs:1751]: I guess we could support @krbz999's suggestion without too much trouble (assuming we relax the `item.system.uses.value !== item.system.uses.max` in the the `if` guard):

```js
const uses = item.system.uses;
const newValue = Math.clamped(uses.value + roll.total, 0, uses.max);
if ( newValue !== uses.value ) {
  const diff = newValue - uses.value;
  const isMax = newValue === uses.max;
  const locKey = `DND5E.Item${diff < 0 ? "Loss" : "Recovery"}Roll${isMax ? "Max": ""}`;
  updates.push({_id: item.id, "system.uses.value": newValue});
  roll.options.flavor = game.i18n.format(locKey, {name: item.name, count: diff});
  rolls.push(roll);
  await roll.toMessage({
    user: game.user.id,
    speaker: {actor: this, alias: this.name}
  });
}
```

**Review by Fyorl — CHANGES_REQUESTED** [review]
Thanks for putting this together, Jeff, this will be a nice addition to the system.

File: module/documents/actor/actor.mjs (hunk 1 above)
Comment (krbz999) [reply module/documents/actor/actor.mjs:1751]: I would also like to suggest that the 7 (for gritty) is not hard coded and instead moved to the CONFIG and able to be overridden.

File: module/documents/actor/actor.mjs [hunk 2]
```diff
@@ -1716,6 +1726,31 @@ export default class Actor5e extends Actor {
       if ( recoverLongRestUses && item.system.recharge?.value ) {
         updates.push({_id: item.id, "system.recharge.charged": true});
       }
+
+      // Items that roll to gain charges on a new day
+      if ( recoverDailyUses && (item.system.uses?.per === "charges")
+           && (item.system.uses.recoveryFormula !== "")
+           && (item.system.uses.value !== item.system.uses.max) ) {
+        const roll = new CONFIG.Dice.DamageRoll(item.system.uses.recoveryFormula, this.getRollData());
```
[the remark this answers is not in the document]
Comment (arbron) [reply module/documents/actor/actor.mjs:1734]: Yeah, that would be a good change.

File: template.json [hunk 3]
```diff
@@ -555,7 +555,8 @@
         "uses": {
           "value": null,
           "max": "",
-          "per": null
+          "per": null,
+          "recoveryFormula": ""
```
[the remark this answers is not in the document]
Comment (arbron) [reply template.json:559]: I think either would work

**Comment by arbron** [comment]
@Fyorl I have no issue with any of your suggestions if you want to make those

**Comment by arbron** [comment]
Here are the compendium changes:

```javascript
[
  {_id: "18fbyArtidKzON01", "data.uses.recoveryFormula": "1d3"},
  {_id: "1RwJWOAeyoideLKe", "data.uses.recoveryFormula": "1d4 + 3"},
  {_id: "1taRIMF9w7jpnonN", "data.uses.recoveryFormula": "1d6 + 1"},
  {_id: "2wxqnjpnPmkpPCC5", "data.uses.recoveryFormula": "1d3"},
  {_id: "4sR5HOah6KwVPHOb", "data.uses.recoveryFormula": "1d6 + 1"},
  {_id: "5Rxo4K9cgpwgW9vZ", "data.uses.recoveryFormula": "1d6 + 1"},
  {_id: "6ai1pEde3iQX30Fr", "data.uses.recoveryFormula": "1d3"},
  {_id: "7jEKkA9qbwJ3IuCb", "data.uses.recoveryFormula": "1d6 + 1"},
  {_id: "DEQkJiQdGyfmSNkV", "data.uses.recoveryFormula": "1d3"},
  {_id: "DoSvjhRARhRqWZXg", "data.uses.recoveryFormula": "1d6 + 1"},
  {_id: "EJFql4aNWHHJSxT9", "data.uses.recoveryFormula": "1d6 + 1"},
  {_id: "ElLfmohtIFMagr5f", "data.uses.recoveryFormula": "1d3"},
  {_id: "FeouSUPUlUhfgeRp", "data.uses.recoveryFormula": "1d8 + 2"},
  {_id: "GcpXNc4dKUNw0Tk6", "data.uses.recoveryFormula": "1d6 + 1"},
  {_id: "HnIERWmmra74hSCw", "data.uses.recoveryFormula": "1d4 + 1"},
  {_id: "JRJpKZyamkpa7awv", "data.uses.recoveryFormula": "1d3"},
  {_id: "L9KBLub5vfb3mTDz", "data.uses.recoveryFormula": "1d4 + 1"},
  {_id: "URun3vYrXKJJdAJe", "data.uses.recoveryFormula": "1d6 + 4"},
  {_id: "VQ6NWjWV37wiB29O", "data.uses.recoveryFormula": "1d3"},
  {_id: "WLVQJVpCWiPkCAtZ", "data.uses.recoveryFormula": "1d6 + 4"},
  {_id: "XZWHQ20ynJBK6xmU", "data.uses.recoveryFormula": "1d4 + 1"},
  {_id: "Zf7kBZa5f4WBepn1", "data.uses.recoveryFormula": "1d4 + 1"},
  {_id: "b46t42bMruQf9v3O", "data.uses.recoveryFormula": "1d3"},
  {_id: "bPFVfq81EsMNu6OQ", "data.uses.recoveryFormula": "1d6 + 1"},
  {_id: "bod1dKzbAkAm21Ho", "data.uses.recoveryFormula": "1d6 + 4"},
  {_id: "caEn3ixCUFBnHTx6", "data.uses.recoveryFormula": "1d6 + 4"},
  {_id: "eM5gEe4SEOvA2Y9t", "data.uses.recoveryFormula": "2d8 + 4"},
  {_id: "etA2SBgWwduCLgLT", "data.uses.recoveryFormula": "1d6"},
  {_id: "jJU8vFhHLQeKe2wu", "data.uses.recoveryFormula": "1d3"},
  {_id: "khyjT3dKyoEOf4eA", "data.uses.recoveryFormula": "1d6 + 1"},
  {_id: "l3V7V8VCXpmAAysQ", "data.uses.recoveryFormula": "4d6 + 2"},
  {_id: "lPsueMv4ZoXqCYf9", "data.uses.recoveryFormula": "1d6 + 1"},
  {_id: "lsiR1hVfISlC5YoB", "data.uses.recoveryFormula": "1d6 + 4"},
  {_id: "nvhk1quD0Dg1ZtSH", "data.uses.recoveryFormula": "1d3"},
  {_id: "o4Irx3hHiD3FnPbL", "data.uses.recoveryFormula": "1d3"},
  {_id: "oNLfJNRQgUHpU8c7", "data.uses.recoveryFormula": "1d3"},
  {_id: "qw05Om9XWqTMoio2", "data.uses.recoveryFormula": "1d3"},
  {_id: "s4fR8bxQGSt4wbH7", "data.uses.recoveryFormula": "1d20"},
  {_id: "skoUe223EvRYGPL6", "data.uses.recoveryFormula": "1d3"},
  {_id: "uHL99JKLUpTKAbz8", "data.uses.recoveryFormula": "1d3"},
  {_id: "wGDDt17DpBcXPuUD", "data.uses.recoveryFormula": "1d4 + 1"},
  {_id: "xEtBeZjJnkDXojQM", "data.uses.recoveryFormula": "1d6 + 4"}
]
```

File: template.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (Varriount) [reply template.json:559]: I would prefer "recoveryFormula"

**Pull request merged**